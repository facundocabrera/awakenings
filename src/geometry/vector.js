const { sqrt, pow, atan, acos } = Math;

// distancia entre 2 puntos en el plano
const distance = (x1, y1, x2, y2) => {
  const d = sqrt(pow(x2 - x1, 2) + pow(y2 - y1, 2));

  return d;
};

// producto scalar entre 2 vectores
const dot_product = (x1, y1, x2, y2) => {
  const dp = x1 * x2 + y1 * y2;

  return dp;
};

// modulo de un vector
const modulo = (x, y) => {
  const d = distance(0, 0, x, y);

  return d;
};

// angulo entre 2 vectores
const angle_between = (x1, y1, x2, y2) => {
  const dp = dot_product(x1, y1, x2, y2);
  const m1 = modulo(x1, y1);
  const m2 = modulo(x2, y2);

  return acos(dp / (m1 * m2));
};

// la direccion de un vector esta dada por el angulo que forma la recta
// que une los 2 puntos y el eje X.
const direction = (x1, y1, x2, y2) => {
  const dy = y2 - y1;
  const dx = x2 - x1;

  return atan(dy / dx);
};

// Con esto traslado los puntos a un vector de radio 1, que me da la
// proporsion que necesito para marcar los puntos en la recta.
const unit_vector = (x, y) => {
  const m = modulo(x, y);

  return [x / m, y / m];
};

const middle_vector = (x1, y1, x2, y2) => {
  return [(x1 + x2) / 2, (y1 + y2) / 2];
};

const sum = ([x, y], [z, t]) => [x + z, y + t];

const sumAll = (center, all) => all.map((v) => sum(v, center));

const mult = (vector, v) => vector.map((scalar) => scalar * v);

const mult_vector = (space2d, scalar) =>
  space2d.map(([x, y]) => [x * scalar, y * scalar]);

const sum_vector = (vector, init = [0, 0]) => vector.reduce(sum, [0, 0]);

const div = (vector, by) => vector.map((v) => v / by);

const centroid = (points) => div(sum_vector(points), points.length);

export {
  distance,
  dot_product,
  modulo,
  angle_between,
  direction,
  unit_vector,
  middle_vector,
  centroid,
  sum,
  sumAll,
  mult,
  mult_vector,
};
