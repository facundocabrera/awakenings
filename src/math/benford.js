import range from "lodash/range";
import log from "./log";

const cumulative = base => [
  ...range(2, base).map(v => log(v, base)),
  1
];

const distribution = base => range(1, base).map(v => log(1 + 1 / v, base));

export {
  cumulative,
  distribution
};