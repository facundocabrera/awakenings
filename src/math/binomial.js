// https://www.khanacademy.org/math/statistics-probability/random-variables-stats-library/binomial-random-variables/v/probability-of-making-2-shots-in-6-attempts

import { pow, combinations } from "mathjs";
import range from "lodash/range";
import { cumulative } from "./stats";

/**
 * Binomial Mass.
 *
 * @param {Number} n independent trials
 * @param {Number} k number of times for a specific outcome within n trials
 * @param {Number} p successes probability for a single trial
 *
 * @returns Number between [0, 1]
 */
const mass = (n, k, p) => combinations(n, k) * pow(p, k) * pow(1 - p, n - k);

/**
 * Binomial distribution.
 *
 * Comienzo desde 0 exitos, hasta evaluar n exitos en los n intentos.
 *
 * @param {Number} n independent trials
 * @param {Number} p successes probability for a single trial
 */
const distribution = (n, p) => range(0, n).map((s) => mass(n, s, p));

export { mass, distribution, cumulative };
