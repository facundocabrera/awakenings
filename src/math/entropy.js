import log from "./log";

/**
 * Self-information of an event.
 * 
 * @param {Number} probability of the event
 * @param {Number} base unit of self-information
 */
const I = (probability, base) => -1 * log(probability, base);

/**
 * Entropy.
 * 
 * Se puede interpretar como el valor esperado para un conjunto de 
 * self-informative events. Basicamente es la mean usando la funcion I.
 * 
 * @param {Array} distribution probabilities for each different event
 * @param {Number} base base to be used in the log calculation
 * 
 * @returns entropy Number between [0, 1]
 */
const H = (distribution, base = 10) => distribution.reduce(
  (h, p) => h -= p * log(p, base), 0
);

export {
  I,
  H
};