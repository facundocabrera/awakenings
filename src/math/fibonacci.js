function* fibonacci(n) {
  const infinite = !n && n !== 0;
  let current = 1;
  let next = 1;

  while (infinite || n--) {
    yield current;
    [current, next] = [next, current + next];
  }
}

function sequenceOf(length) {
  const s = [];
  const gen = fibonacci(length);

  for(let v of gen) {
    s.push(v);
  }

  return s;
}

function ratios(accuracy = 27, nFollowing = 9) {
  const s = sequenceOf(accuracy);
  const r = [];
  let last = accuracy - 1;
  let previous = last - 1;

  while(last - previous <= nFollowing) {
    r.push(s[previous] / s[last]);
    previous--;
  }

  return r;
}

function cumulative(n) {
  if (n < 5)
    console.warn('are you sure?');

  const r = ratios(144, n).map(v => 1 - v);

  return r;
}

export {
  fibonacci,
  sequenceOf,
  ratios,
  cumulative
};