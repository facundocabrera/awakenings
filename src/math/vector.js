import { range } from "lodash";
import MersenneTwister from "mersenne-twister";

const generator = new MersenneTwister();

export const random_vector = (size) => {
  const vector = [];

  for (let i = 0; i < size; i++) {
    // le saco precision porque no la necesito para lo que estoy simulando
    vector.push(generator.random().toFixed(3));
  }

  return vector;
};

export const fixed_vector = (size, value) => {
  const vector = [];

  for (let i = 0; i < size; i++) {
    vector.push(value);
  }

  return vector;
};

export const seq_vector = (size, step = 1, shift = 0) =>
  range(shift, size * step + shift, step);
