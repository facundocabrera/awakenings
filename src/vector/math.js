import { zip } from "lodash";

/**
 * https://en.wikipedia.org/wiki/Dot_product
 * Algebraic definition
 */
const dot = (vector1, vector2) => {
  if (vector1.length === vector2.length) {
    return zip(vector1, vector2)
      .map(([v1, v2]) => v1 * v2)
      .reduce((sum, n) => sum + n, 0);
  }

  throw Error("dot / lengths must be equal");
};

const scale2D = (space2d, weights) => {
  if (space2d.length === weights.length) {
    return zip(space2d, weights).map(([[x, y], w]) => [x * w, y * w]);
  }

  throw Error("scale2D / lengths should be equals");
};

const scale3D = (space3d, weights) =>
  zip(space3d, weights).map(([[x, y, z], w]) => [x * w, y * w, z * w]);

const scaleND = (spaceNd, weights) =>
  zip(spaceNd, weights).map(([vector, w]) => vector.map((v) => v * w));

export { dot, scale2D, scale3D, scaleND };
