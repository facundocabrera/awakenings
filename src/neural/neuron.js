import { zip } from "lodash";

import { stops } from "../geometry/circle";
import { centroid } from "../geometry/vector";
import { random_vector } from "../math/vector";

export const apply = (points, weights) =>
  zip(points, weights).map(([[x, y], w]) => [x * w, y * w]);

export const Neuron = ({ connections }) => {
  // puntos alrededor de un circulo (vectores unitarios).
  const vector = stops(connections);

  // vector aleatorio de weights para simular comportamiento.
  const weights = random_vector(connections);

  // vectores weights
  const wv = apply(vector, weights);

  // activar neurona con vector de entrada
  const activate = (input) => centroid(apply(wv, input));

  return {
    activate,
  };
};

export const TrackNeuronActivationDomain = (neuron) => {
  const { activate } = neuron;
  let min = [Infinity, Infinity];
  let max = [-Infinity, -Infinity];

  const tracker = (...args) => {
    const output = activate(...args);

    console.log(output);

    min = [Math.min(min[0], output[0]), Math.min(min[1], output[1])];
    max = [Math.max(max[0], output[0]), Math.max(max[1], output[1])];
  };

  const box = () => [min, max];

  return {
    ...neuron,
    box,
    activate: tracker,
  };
};
