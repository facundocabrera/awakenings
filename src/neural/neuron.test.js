import { random_vector } from "../math/vector";
import { apply, Neuron, TrackNeuronActivationDomain } from "./neuron";

jest.mock("../math/vector");

beforeEach(() => {
  random_vector.mockImplementation((n) => Array(n).fill(1));
});

describe("Neuron", () => {
  it("should activate", () => {
    const n = Neuron({ connections: 3 });

    expect(n.activate([1, 2, 3])).toBeDefined();
    expect(n.activate([1, 2, 3])).toMatchSnapshot();
  });

  it("should expose internals", () => {
    const n = Neuron({ connections: 5 });

    expect(n.internals).toMatchSnapshot();
  });
});

describe("TrackNeuronActivationDomain", () => {
  it("should calculate min and max", () => {
    const n = TrackNeuronActivationDomain(Neuron({ connections: 3 }));

    n.activate([1, 2, 3]);
    n.activate([3, 2, 1]);

    expect(n.box()).toMatchSnapshot();
  });
});

describe("apply", () => {
  it("should scale the input by the weights", () => {
    expect(
      apply(
        [
          [1, 2], // * 0.3
          [3, 4], // * 0.7
        ],
        [0.3, 0.7]
      )
    ).toMatchSnapshot();
  });
});
