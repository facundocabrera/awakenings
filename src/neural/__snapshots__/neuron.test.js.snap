// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Neuron should calculate min and max 1`] = `
Array [
  -0.5,
  -0.2886666666666666,
]
`;

exports[`apply should scale the input by the weights 1`] = `
Array [
  Array [
    0.3,
    0.6,
  ],
  Array [
    2.0999999999999996,
    2.8,
  ],
]
`;
